### DISCUSSION TITLE
Path of least resistance to write CRTP in TypeScript

### DISCUSSION THREAD

**Discussion by dbaldwin** [bitcraze/discussions#962] [General]
A couple of years ago I began working on understanding CRTP and writing a proof of concept together in TypeScript (NodeJS) via Bluetooth. I've been revisiting the project recently as I think it would able a much cleaner/simpler way to get into the world of Crazyflie. Especially for classrooms that may not have the knowledge or equipment to run Python. I know this sounds crazy, but we run into it every day with our user base.

That being said, what is the easiest way to understand the raw data being sent from cflib to Crazyflie? I've been digging through the [code here](https://github.com/bitcraze/crazyflie-lib-python/tree/master/cflib/crtp), but curious what others think. I'd like to understand the raw packets I'd need to send to takeoff, fly forward 1 meter, and land. This is very simple using the motion commander in Python, but I want the raw packets so I can do a TypeScript implementation. Any thoughts/suggestions would be very much appreciated. Thanks in advance.

**Comment by knmcguire** [comment]
Hi!

Let me give you some links to get started:

- The CRTP documentation should tell you the actual content of each packet: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/crtp/
- To append that, you can also look at the crtp decoding on the crazyflie's firmware (anything that starts with crtp_*.c ): https://github.com/bitcraze/crazyflie-firmware/tree/master/src/modules/src
- There have been quite some people that wrote typescript for the crazyflie crtp as an external project. They are quite old but perhaps could use as a reference: https://www.bitcraze.io/development/external-projects/#connectivity-libraries

Hope this helps as a start! If you need any help in the process about anything specific, just start a new discussion :)

**Reply by dbaldwin** [reply]
Great info, @knmcguire. Will study up. Thank you.
